api/src/main/java/io/strimzi/api/kafka/model/listener/IngressListenerBootstrapConfiguration.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2018, Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.api.kafka.model.listener;
+
+import com.fasterxml.jackson.annotation.JsonInclude;
+import io.strimzi.crdgenerator.annotations.Description;
+import io.sundr.builder.annotations.Buildable;
+import lombok.EqualsAndHashCode;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Configures Ingress for Bootstrap service
+ */
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@Buildable(
+    editableEnabled = false,
+    generateBuilderPackage = false,
+    builderPackage = "io.fabric8.kubernetes.api.builder"
+)
+@EqualsAndHashCode(callSuper = true)
+public class IngressListenerBootstrapConfiguration extends ExternalListenerBootstrapOverride {
+    private static final long serialVersionUID = 1L;
+
+    private String host;
+    private Map<String, String> dnsAnnotations = new HashMap<>(0);
+
+    @Description("Host for the bootstrap route. " +
+            "This field will be used in the Ingress resource.")
+    public String getHost() {
+        return host;
+    }
+
+    public void setHost(String host) {
+        this.host = host;
+    }
+
+    @Description("Annotations which will be added to the Ingress resource. " +
+            "You can use this field to instrument DNS providers such as External DNS.")
+    @JsonInclude(JsonInclude.Include.NON_EMPTY)
+    public Map<String, String> getDnsAnnotations() {
```

Comment:
They these annotation are added to the `Ingress` why is this not called `getIngressAnnotations()`?

If we're supporting annotations on the ingress why not labels?

Reply:
It is named `dnsAnnotations` and it supports only annotations because of the particular way how DNS is orchestrated when using Ingress in combination with tools like ExternalDNS. You attach some special annotations and the ExternalDNS tooling will automatically create the corresponding DNS records for example in Google CloudDNS or Amazon Route53. This is done through annotations and not labels.

So for example following YAML:

```yaml
# ...
listeners:
  external:
    type: ingress
    authentication:
      type: tls
    configuration:
      bootstrap:
        dnsAnnotations:
          external-dns.alpha.kubernetes.io/hostname: bootstrap.j9z.cz.
          external-dns.alpha.kubernetes.io/ttl: "60"
        host: bootstrap.j9z.cz
      brokers:
      - broker: 0
        dnsAnnotations:
          external-dns.alpha.kubernetes.io/hostname: broker-0.j9z.cz.
          external-dns.alpha.kubernetes.io/ttl: "60"
        host: broker-0.j9z.cz
      - broker: 1
        dnsAnnotations:
          external-dns.alpha.kubernetes.io/hostname: broker-1.j9z.cz.
          external-dns.alpha.kubernetes.io/ttl: "60"
        host: broker-1.j9z.cz
      - broker: 2
        dnsAnnotations:
          external-dns.alpha.kubernetes.io/hostname: broker-2.j9z.cz.
          external-dns.alpha.kubernetes.io/ttl: "60"
        host: broker-2.j9z.cz
# ...
```

Would in my Kubernetes cluster create the Ingress resources and automatically create the DNS records to route the traffic to the Ingress.

Reply:
That said, if you want, I can make it more general. But it was really created with this purpose in mind.

Reply:
I see now that you support annotations and labels on the ingress via the template mechanism, which answers my question about labels. These dnsAnnotations mean there are two places to configure the annotations on the `Ingress`. I understand your motivation to provide all the info necessary to use the ingress in one place, but having two places to configure something is not great.

Reply:
There is one big difference between what is in the template and what is here. Here you can set them up per broker (required here, because the DNS records need to be different for each Kafka broker). In templates you just set a general annotations for all broker ingress resources at once. So that cannot be used. 

I have thought about this originally. But as I said, this is really a single purpose field to orchestrate the DNS. That is why I think it belongs a bit better here, because here you configure the hostnames for the ingresses. If you move this to templates, it will mean that you have to configure these hostnames on two different places and not just in one.

I also didn't liked the spreading of the pattern with the broker IDs into the template field (although maybe one day we will anyway get there).